repository_service_tuf/cli/admin/metadata.py

```diff
@@ -0,0 +1,605 @@
+        ]
+        keys_table = _create_keys_table(signing_list, current_root, False)
+        console.print("\nHere are the current root signing keys:")
+        console.print(keys_table)
+        keys_amount = len(current_root.signing_keys)
+        if root_threshold <= keys_amount:
+            agree = prompt.Confirm.ask("\nDo you want to add a new key?")
+            if not agree:
+                return
+        else:
+            remaining = root_threshold - keys_amount
+            console.print(f"You must add {remaining} more signing key(s)")
+
+        root_key: RSTUFKey = _get_key(Root.type)
+        if root_key.error:
+            console.print(f":cross_mark: [red]Failed[/]: {root_key.error}")
+            continue
+
+        if root_key == current_root.online_key:
+            console.print(
+                ":cross_mark: [red]Failed[/]: This is the current online key. "
+                "Cannot be added"
+            )
+            continue
+
+        if current_root.is_keyid_used(root_key.key["keyid"]):
+            console.print(":cross_mark: [red]Failed[/]: Key is already used")
+            continue
+
+        root_key.name = prompt.Prompt.ask(
+            "[Optional] Give a [green]name/tag[/] to the key"
+        )
+
+        current_root.add_key(root_key)
+
+
+def _get_positive_int_input(msg: str, input_name: str, default: Any) -> int:
+    input: int = 0
+    while True:
+        input = prompt.IntPrompt.ask(msg, default=default, show_default=True)
+        if input < 1:
+            console.print(f"{input_name} must be at least 1")
+        else:
+            break
+
+    return input
+
+
+def _modify_expiration(current_root: RootInfo):
+    console.print(markdown.Markdown(EXPIRY_CHANGES_MSG), width=100)
+    console.print("\n")
+    while True:
+        console.print(
+            f"Current root expiration: [cyan]{current_root.expiration_str}[/]",
+            highlight=False,  # disable built-in rich highlight
+        )
+        if current_root.expiration < (datetime.now() + timedelta(days=1)):
+            console.print(
+                "You must extend root's expiration - root has expired"
+            )
+            break
+
+        change = prompt.Confirm.ask(
+            "Do you want to extend the [cyan]root's expiration[/]?"
+        )
+        if not change:
+            console.print("Skipping root expiration changes")
+            return
+        else:
+            break
+
+    msg = "Days to extend [cyan]root's expiration[/] starting from today"
+    while True:
+        expiry_bump = _get_positive_int_input(msg, "Expiration extension", 365)
+        new_expiry = datetime.now() + timedelta(days=expiry_bump)
+        new_exp_str = new_expiry.strftime("%Y-%b-%d")
+        agree = prompt.Confirm.ask(
+            f"New root expiration: [cyan]{new_exp_str}[/]. Do you agree?"
+        )
+        if agree:
+            current_root.expiration = new_expiry
+            return
```

Comment:
nit: I don't think you need two loops here. what about something like...

```python
while True:

    if root.is_expired:  # this may already exist in Metadata API, if not you could add it to RootInfo
        print("root expired. please change!")
        change = True

    else:
        change = prompt(f"root expires on {root.expiration}. change? y/n") # expiration getter could do the formatting

    if not change:
        break

    root.expiration = prompt("enter days from now") # setter could do the math
```

Reply:
This is roughly the same flow, which you have in `_keys_additions`, and which I think is pretty good. 👍

Reply:
I think I found a way to simplify it to one cycle.